// publish-please/lib/validations/sensitive-data.js
'use strict';

var isSensitiveData = require('ban-sensitive-files');
var globby = require('globby');
var confirm = require('../utils/inquires').confirm;
var inputList = require('../utils/inquires').inputList;
var Promise = require('pinkie-promise');

module.exports = {
    option: 'sensitiveData',
    statusText: 'Checking for the sensitive data in the working tree',
    defaultParam: true,

    configurator: function configurator(currentVal) {
        function configureIgnores() {
            var ignore = Array.isArray(currentVal.ignore) ? currentVal.ignore : [];

            return confirm('Is there any files that you want to exclude from check?', false).then(function (yes) {
                return yes ? inputList('List files you want to exclude (comma-separated, you can use glob patterns)', ignore) : true;
            }).then(function (answer) {
                return Array.isArray(answer) ? { ignore: answer } : answer;
            });
        }

        return confirm('Would you like to verify that there is no sensitive data in your working tree?', !!currentVal).then(function (yes) {
            return yes ? configureIgnores() : false;
        });
    },
    run: function run(opts, pkgInfo) {
        return Promise.resolve().then(function () {
            if (opts && Array.isArray(opts.ignore)) return globby(opts.ignore);

            return [];
        }).then(function (ignore) {
            var errs = [];
            var addErr = errs.push.bind(errs);

            pkgInfo.files.filter(function (path) {
                return ignore.indexOf(path) < 0;
            }).forEach(function (path) {
                return isSensitiveData(path, addErr);
            });

            if (errs.length) {
                var msg = errs.map(function (err) {
                    return err.split(/\n/).map(function (line) {
                        return '    ' + line;
                    }).join('\n');
                }).join('\n');

                throw 'Sensitive data found in the working tree:\n' + msg;
            }
        });
    }
};
