// publish-please/lib/utils/inquires.js
'use strict';

var inquirer = require('inquirer');
var Promise = require('pinkie-promise');

function ask(type, question, defaultAnswer) {
    return new Promise(function (resolve) {
        inquirer.prompt({
            type: type,
            name: 'prop',
            message: question,
            default: defaultAnswer
        }, function (answer) {
            return resolve(answer['prop']);
        });
    });
}

var confirm = exports.confirm = function (question, defaultAnswer) {
    return ask('confirm', question, defaultAnswer);
};

var input = exports.input = function (question, defaultAnswer) {
    return ask('input', question, defaultAnswer);
};

exports.inputWithConfirmation = function (confirmQuestion, denialAnswer, inputQuestion, defaultAnswer) {
    return confirm(confirmQuestion, true).then(function (yes) {
        return yes ? input(inputQuestion, defaultAnswer) : denialAnswer;
    });
};

exports.inputList = function (question, defaultAnswer) {
    return input(question, defaultAnswer).then(function (answer) {
        return Array.isArray(answer) ? answer : answer.split(/,\s*/);
    });
};
