// publish-please/lib/publish.js
'use strict';

var readFile = require('fs').readFileSync;
var chalk = require('chalk');
var semver = require('semver');
var defaults = require('lodash/defaultsDeep');
var pkgd = require('pkgd');
var exec = require('cp-sugar').exec;
var spawn = require('cp-sugar').spawn;
var emoji = require('node-emoji').emoji;
var validate = require('./validations').validate;
var confirm = require('./utils/inquires').confirm;
var DEFAULT_OPTIONS = require('./default-options');

var SCRIPT_TYPE = {
    prePublish: 'pre-publish',
    postPublish: 'post-publish'
};

function printReleaseInfo(pkgVersion, publishTag) {
    var commitInfo = null;

    return exec('git log -1 --oneline').then(function (info) {
        commitInfo = info;

        return exec('npm whoami --silent');
    }).catch(function () {
        return chalk.red('<not logged in>');
    }).then(function (publisher) {
        console.log(chalk.yellow('Release info'));
        console.log(chalk.yellow('------------'));
        console.log('  ' + chalk.magenta('Version:       ') + pkgVersion);
        console.log('  ' + chalk.magenta('Latest commit: ') + commitInfo);
        console.log('  ' + chalk.magenta('Publish tag:   ') + publishTag);
        console.log('  ' + chalk.magenta('Publisher:     ') + publisher);
        console.log();
    });
}

function runScript(command, scriptType) {
    if (!module.exports.testMode) {
        console.log(chalk.yellow('Running ' + (scriptType ? scriptType + ' ' : '') + 'script'));
        console.log(chalk.yellow('-------------------------'));
    }

    return spawn(command, module.exports.testMode).then(function () {
        if (!module.exports.testMode) {
            console.log(chalk.yellow('-------------------------'));
            console.log(emoji['+1'], emoji['+1'], emoji['+1']);
            console.log();
        }
    });
}

function publish(publishTag) {
    var command = 'npm publish --tag ' + publishTag + ' --with-publish-please';

    if (!module.exports.testMode) return spawn(command).then(function () {
        return console.log('\n', emoji.tada, emoji.tada, emoji.tada);
    });

    return command;
}

function getOptions(opts) {
    var rcFileContent = null;
    var rcOpts = {};

    try {
        rcFileContent = readFile('.publishrc').toString();
    } catch (err) {
        // NOTE: we don't have .publishrc file, just ignore the error
    }

    if (rcFileContent) {
        try {
            rcOpts = JSON.parse(rcFileContent);
        } catch (err) {
            throw new Error('.publishrc is not a valid JSON file.');
        }

        opts = defaults({}, opts, rcOpts);
    }

    return defaults({}, opts, DEFAULT_OPTIONS);
}

// NOTE: adopted from https://github.com/sindresorhus/np/blob/master/index.js#L78
function assertNode6PublishingPrerequisites() {
    return exec('npm version --json').then(function (out) {
        var npmVersion = JSON.parse(out).npm;
        var isNode6 = semver.gte(process.version, '6.0.0');
        var isSafeNpmVersion = semver.satisfies(npmVersion, '>=2.15.8 <3.0.0 || >=3.10.1');

        if (isNode6 && !isSafeNpmVersion) throw new Error('npm@' + npmVersion + ' has known issues publishing when running Node.js 6. Please upgrade npm or downgrade Node and publish again. See: https://github.com/npm/npm/issues/5082');
    });
}

module.exports = function (opts) {
    var pkgInfo = null;

    opts = getOptions(opts);

    return assertNode6PublishingPrerequisites().then(function () {
        return opts.prePublishScript && runScript(opts.prePublishScript, SCRIPT_TYPE.prePublish);
    }).then(function () {
        return pkgd();
    }).then(function (info) {
        return pkgInfo = info;
    }).then(function () {
        return validate(opts.validations, pkgInfo);
    }).then(function () {
        return !module.exports.testMode && printReleaseInfo(pkgInfo.cfg.version, opts.publishTag);
    }).then(function () {
        return opts.confirm ? confirm('Are you sure you want to publish this version to npm?', false) : true;
    }).then(function (ok) {
        return ok && publish(opts.publishTag) || '';
    }).then(function (command) {
        if (!command || !opts.postPublishScript) return command;

        return runScript(opts.postPublishScript, SCRIPT_TYPE.postPublish).then(function () {
            return command;
        });
    });
};

// Exports for the testing purposes
module.exports.testMode = false;
module.exports.getOptions = getOptions;
